internal/metrics/v2/gauge_test.go

```diff
@@ -0,0 +1,447 @@
+					Name:       "gauge_test_total",
+					Help:       "gauge_test help",
+					LabelNames: []string{"foo", "bar"},
+				},
+			},
+			args: args{
+				label: []Label{{"foo", "fooval"}, {"bar", "barval"}},
+			},
+		},
+	}
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			g := NewGauge(test.fields.registry, test.fields.opts)
+			g.Set(test.args.floatVal, test.args.label...)
+			family, err := test.fields.registry.Gather()
+			require.NoError(t, err)
+			require.Equal(t, test.expect, family[0].Metric[0].Gauge.GetValue())
+			require.Equal(t, len(test.fields.opts.LabelNames), len(family[0].Metric[0].GetLabel()))
+		})
+	}
+}
+
+func TestPrometheusGaugeSub(t *testing.T) {
+	type fields struct {
+		registry *prometheus.Registry
+		opts     GaugeOpts
+	}
+	type args struct {
+		floatVal float64
+		label    []Label
+	}
+	tests := []struct {
+		name   string
+		fields fields
+		args   args
+		expect float64
+	}{
+		{
+			name: "should subtract",
+			fields: fields{
+				registry: prometheus.NewRegistry(),
+				opts: GaugeOpts{
+					Subsystem:  "cp",
+					Name:       "gauge_test_total",
+					Help:       "gauge_test help",
+					LabelNames: []string{},
+				},
+			},
+			args: args{
+				floatVal: float64(1),
+				label:    []Label{},
+			},
+			expect: float64(-1),
+		},
+		{
+			name: "should subtract with labels",
+			fields: fields{
+				registry: prometheus.NewRegistry(),
+				opts: GaugeOpts{
+					Subsystem:  "cp",
+					Name:       "gauge_test_total",
+					Help:       "gauge_test help",
+					LabelNames: []string{"foo", "bar"},
+				},
+			},
+			args: args{
+				floatVal: float64(20),
+				label:    []Label{{"foo", "fooval"}, {"bar", "barval"}},
+			},
+			expect: float64(-20),
+		},
+		{
+			name: "should behave like add when input is < 0",
+			fields: fields{
+				registry: prometheus.NewRegistry(),
+				opts: GaugeOpts{
+					Subsystem:  "cp",
+					Name:       "gauge_test_total",
+					Help:       "gauge_test help",
+					LabelNames: []string{"foo", "bar"},
+				},
+			},
+			args: args{
+				floatVal: float64(-2),
+				label:    []Label{{"foo", "fooval"}, {"bar", "barval"}},
+			},
+			expect: float64(2),
+		},
+		{
+			name: "should succeed when when input is 0",
+			fields: fields{
+				registry: prometheus.NewRegistry(),
+				opts: GaugeOpts{
+					Subsystem:  "cp",
+					Name:       "gauge_test_total",
+					Help:       "gauge_test help",
+					LabelNames: []string{"foo", "bar"},
+				},
+			},
+			args: args{
+				label: []Label{{"foo", "fooval"}, {"bar", "barval"}},
+			},
+		},
+	}
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			g := NewGauge(test.fields.registry, test.fields.opts)
+			g.Sub(test.args.floatVal, test.args.label...)
+			family, err := test.fields.registry.Gather()
+			require.NoError(t, err)
+			require.Equal(t, test.expect, family[0].Metric[0].Gauge.GetValue())
+			require.Equal(t, len(test.fields.opts.LabelNames), len(family[0].Metric[0].GetLabel()))
```

Comment:
`requre.Len()` exists for this (and the other places where this is being done) 😄

Reply:
[cc5d6d7](https://github.com/Kong/koko/pull/364/commits/cc5d6d7753178dfab3aaab59957ddd8a53c0650c)